streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowedCogroupedKStreamImplTest.java

```diff
@@ -0,0 +1,271 @@
+        cogroupedStream = groupedStream.cogroup(MockAggregator.TOSTRING_ADDER).cogroup(groupedStream2, MockAggregator.TOSTRING_REMOVER);
+        windowedCogroupedStream = cogroupedStream.windowedBy(TimeWindows.of(ofMillis(500L)));
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullInitializerOnAggregate() {
+        windowedCogroupedStream.aggregate(null);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullMaterializedOnTwoOptionAggregate() {
+        windowedCogroupedStream.aggregate(MockInitializer.STRING_INIT, (Materialized<String, String, WindowStore<Bytes, byte[]>>) null);
+    }
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullNamedTwoOptionOnAggregate() {
+        windowedCogroupedStream.aggregate(MockInitializer.STRING_INIT, (Named) null);
+    }
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullInitializerTwoOptionNamedOnAggregate() {
+        windowedCogroupedStream.aggregate(null, Named.as("test"));
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullInitializerTwoOptionMaterializedOnAggregate() {
+        windowedCogroupedStream.aggregate(null, Materialized.as("test"));
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullInitializerThreeOptionOnAggregate() {
+        windowedCogroupedStream.aggregate(null, Named.as("test"), Materialized.as("test"));
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullMaterializedOnAggregate() {
+        windowedCogroupedStream.aggregate(MockInitializer.STRING_INIT, Named.as("Test"), null);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullNamedOnAggregate() {
+        windowedCogroupedStream.aggregate(MockInitializer.STRING_INIT, null, Materialized.as("test"));
+    }
+
+    @Test
+    public void timeWindowAggregateTest() {
+        final KTable<Windowed<String>, String> customers = groupedStream.cogroup(MockAggregator.TOSTRING_ADDER)
+                .windowedBy(TimeWindows.of(ofMillis(500L))).aggregate(MockInitializer.STRING_INIT, Materialized.with(Serdes.String(), Serdes.String()));
+        customers.toStream().to(OUTPUT);
+
+        try (final TopologyTestDriver driver = new TopologyTestDriver(builder.build(), props)) {
+            final TestInputTopic<String, String> testInputTopic = driver.createInputTopic(TOPIC, new StringSerializer(), new StringSerializer());
+            final TestOutputTopic<Windowed<String>, String> testOutputTopic = driver.createOutputTopic(OUTPUT, new TimeWindowedDeserializer<>(new StringDeserializer()), new StringDeserializer());
+            testInputTopic.pipeInput("k1", "A", 0);
+            testInputTopic.pipeInput("k2", "A", 0);
+            testInputTopic.pipeInput("k2", "A", 1);
+            testInputTopic.pipeInput("k1", "A", 2);
```

Comment:
I think we should pipe some more inputs to verify the window boundaries (ie, ts=499, and ts=500)

We should also add a test for an overlapping window.

Reply:
Just see the test from below with "many windows" -- I think we can merge them into one test